File: week1/보연/BJ13549_숨바꼭질3.cpp
#include <iostream>
#include <cstring>
#include <queue>

using namespace std;

const int MAX_SIZE = 100001;

int simulation(int N, int K) {
	int visited[MAX_SIZE];
	for (int i = 0; i < MAX_SIZE; i++) {
		visited[i] = 2e9;
	}
	queue<int> que;
	que.push(N);
	visited[N] = 0;

	while (!que.empty()) {
		int curr = que.front();
		que.pop();

		if (curr == K) return visited[K];

		// 거리 + 1
		if (curr + 1 < MAX_SIZE && visited[curr + 1] > visited[curr] + 1) {
			visited[curr + 1] = visited[curr] + 1;
			que.push(curr + 1);
		}
		// 거리 - 1
		if (curr - 1 >= 0 && visited[curr - 1] > visited[curr] + 1) {
			visited[curr - 1] = visited[curr] + 1;
			que.push(curr - 1);
		}

		// 순간이동 *2
		if (curr * 2 < MAX_SIZE && visited[curr * 2] > visited[curr]) {
			visited[curr * 2] = visited[curr];
			que.push(curr * 2);
		}
	}
}

int main(){
	ios::sync_with_stdio(false);
	cin.tie(0);
	int N, K;
	//cin >> N >> K;
	N = 5, K = 17;

	int ans = 0;

	cout << simulation(N, K);

	return 0;
}

// import java.util.Arrays;
// import java.util.LinkedList;
// import java.util.Queue;

// public class Main {
//     static final int MAX_SIZE = 100001;

//     static int simulation(int N, int K) {
//         int[] visited = new int[MAX_SIZE];
//         Arrays.fill(visited, 2_000_000_000);
//         Queue<Integer> queue = new LinkedList<>();
//         queue.add(N);
//         visited[N] = 0;

//         while (!queue.isEmpty()) {
//             int curr = queue.poll();

//             if (curr == K) return visited[K];

//             // 거리 + 1
//             if (curr + 1 < MAX_SIZE && visited[curr + 1] > visited[curr] + 1) {
//                 visited[curr + 1] = visited[curr] + 1;
//                 queue.add(curr + 1);
//             }
//             // 거리 - 1
//             if (curr - 1 >= 0 && visited[curr - 1] > visited[curr] + 1) {
//                 visited[curr - 1] = visited[curr] + 1;
//                 queue.add(curr - 1);
//             }

//             // 순간이동 *2
//             if (curr * 2 < MAX_SIZE && visited[curr * 2] > visited[curr]) {
//                 visited[curr * 2] = visited[curr];
//                 queue.add(curr * 2);
//             }
//         }
//         return -1; // 도달할 수 없는 경우
//     }

//     public static void main(String[] args) {
//         int N = 5, K = 17;
//         int ans = simulation(N, K);
//         System.out.println(ans);
//     }
// }

File: week1/지환/BJ13549_숨바꼭질3.java
package week1.지환;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class BJ13549_숨바꼭질3 {
	public static void main(String args[]) throws Exception {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");
		int n = Integer.parseInt(st.nextToken());
		int k = Integer.parseInt(st.nextToken());
		System.out.println(bfs(n, k));
	}

	public static int bfs(int n, int k) {
		int[] visited = new int[100001];
		Arrays.fill(visited, -1);
		Queue<Integer> q = new LinkedList<>();
		q.add(n);
		visited[n] = 0;
		while(!q.isEmpty()) {
			int tmp = q.poll();
			if(tmp == k) {
				return visited[k];
			}

			if(tmp*2 <= 100000 && visited[tmp*2]==-1) {
				visited[tmp*2] = visited[tmp];
				q.add(tmp*2);
			}

			if(tmp-1 >= 0 && visited[tmp-1]==-1) {
				q.add(tmp-1);
				visited[tmp-1] = visited[tmp] + 1;
			}

			if(tmp+1 <= 100000 && visited[tmp+1]==-1) {
				visited[tmp+1] = visited[tmp] + 1;
				q.add(tmp+1);
			}
		}
		return 0;
	}
}
